### PR TITLE
Implemented autoformatting on save

### PR DESCRIPTION
**Pull request by adriantoine**
Hi,

This allows to autoformat on save, and I added some documentation.
Fixes https://github.com/danreeves/sublime-prettier/issues/8

Let me know what you think!

Thanks

### CODE REVIEW

**Comment by adriantoine** [comment]
mmmh actually it doesn't work, looked a bit too easy 🤔

**Comment by danreeves** [comment]
Hey @adriantoine 

Thanks for the the PR. Are you sure it doesn't work because it looks correct to me. I'll try it later and let you know.

I'd love to help you get a PR merged 😀

**Comment by adriantoine** [comment]
Hi @danreeves,

Thanks for your help! Sorry I didn't have time to work on that, the issue is that it was reformatting the code before saving. The plugin was running `prettier` on the unsaved file, which means that it would reformat the code before you changed it, so you would lose all the changes you made to the file.

I don't know if it's clear, I tried `on_post_save` as well, but it wouldn't save the `prettier` changes then. I'd be happy to look into it.

PS: What happened to the issues on this repo?

**Comment by danreeves** [comment]
There's another plugin, but it does things in a way I don't really like. I'll try to take a look at this today. For a start the settings you added to the README should be in JSON, since that's how it's handled by Sublime.

Thanks

File: README.md [hunk 1]
```diff
@@ -35,3 +35,32 @@ Open the command palette and search for `Prettier: Auto format this file`
 ## Configuration
 
 The plugin takes the same settings and the `prettier` tool. See the [`prettier` repo](https://github.com/jlongster/prettier/blob/master/src/options.js) or [this repo](https://github.com/danreeves/sublime-prettier/blob/master/Prettier.sublime-settings). You can configure them in Sublime at `Preferences > Package Settings > Prettier`.
+
+You can turn off the auto formatting on save by setting `autoformat` to `false`.
+
+For example:
+
+```js
```
Comment (danreeves) [inline README.md:43]: These options should be shown in JSON

**Review by danreeves — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (adriantoine) [reply README.md:43]: I agree but the GitHub syntax highlight is being weird with comments and show red boxes, with the JSON syntax.

**Comment by adriantoine** [comment]
Ok thanks, I'll see if I can fix it myself

**Comment by danreeves** [comment]
You'll also need to add the default autoformat setting in `Prettier.sublime-settings`. I would prefer if autoformatting was opt-in, so default to `false`.

**Comment by adriantoine** [comment]
Ok no problem, I'll add it

File: README.md (hunk 1 above)
Comment (danreeves) [reply README.md:43]: Good point. I think it'll be fine if it still uses js highlighting but is JSON. e.g.:

```js
{
  // Comments are allowed in JS
  "printWidth": 80,
  // But not allowed in JSON
  "tabWidth": 2,
  "useFlowParser": false,
  "singleQuote": false,
  "trailingComma": false,
  "bracketSpacing": true
}
```

I _think_ Sublime handles comments anyway. Could be wrong though

File: prettier.py [hunk 2]
```diff
@@ -1,7 +1,10 @@
-from subprocess import Popen, check_output, CalledProcessError, PIPE
+import subprocess
+import platform
```
Comment (adriantoine) [inline prettier.py:2]: I don't why this is shown as a change, it's already like that in your repo: https://github.com/danreeves/sublime-prettier/blob/master/prettier.py

File: prettier.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
...
 from sublime import Region, load_settings
 import sublime_plugin
 
+IS_WINDOWS = platform.system() == 'Windows'
+
```
Comment (adriantoine) [inline prettier.py:7]: Same there

File: prettier.py (hunk 2 above)
Comment (adriantoine) [reply prettier.py:2]: Yeah the issue is just that I didn't merge the latest changes into my fork

**Comment by adriantoine** [comment]
Hey @danreeves,

I have finally found a way to make it work, but I have changed the way the `prettier` command works.

Before, running the `prettier` command was running the `prettier` executable on the current file, which worked fine, except that it couldn't work with `on_pre_save` because it would format the code before saving and the user would lose his unsaved changes.

What I did is that instead of running the `prettier` command on the file, I run it with `--stdin` set to the whole file content, exactly the way it is done for the `prettier_selection` command, except that instead of formatting the selection, I format the whole file. In that way, you can format the whole file, even the unsaved changes.

So I created the `prettify_code` function which formats and replaces a given `Region`.

Let me know what you think!

**Comment by danreeves** [comment]
This is awesome, I had been planning on refactoring the two commands to work the same. Nice work!

I'm going to pull it down now and give it a go

**Comment by adriantoine** [comment]
Great, thanks!

File: README.md [hunk 4]
```diff
@@ -35,3 +35,32 @@ Open the command palette and search for `Prettier: Auto format this file`
...
 The plugin takes the same settings and the `prettier` tool. See the [`prettier` repo](https://github.com/jlongster/prettier/blob/master/src/options.js) or [this repo](https://github.com/danreeves/sublime-prettier/blob/master/Prettier.sublime-settings). You can configure them in Sublime at `Preferences > Package Settings > Prettier`.
+
+You can turn on the auto formatting on save by setting `autoformat` to `true`.
+
+For example:
+
+```js
+{
+  // Turns on/off autoformatting on save
+  autoformat: true,
```
Comment (danreeves) [inline README.md]: Can you add double quotes around the keys here, so it looks like JSON.

File: prettier.py [hunk 5]
```diff
@@ -35,21 +35,30 @@ def report_error(returncode, err=None):
     if err:
         print(err)
 
+# Calls prettier on the given region and replaces code
```
Comment (danreeves) [inline prettier.py:39]: Two empty lines above function defs please. It's not written anywhere yet but the project is following flake8 linting rules.

File: prettier.py [hunk 6]
```diff
@@ -58,15 +67,13 @@ def run(self, edit):
         command = ['prettier'] + options + ['--stdin']
         for region in self.view.sel():
             text = self.view.substr(region)
```
Comment (danreeves) [inline prettier.py]: `text` isn't used here so this line can be removed

File: prettier.py [hunk 7]
```diff
@@ -58,15 +67,13 @@ def run(self, edit):
         command = ['prettier'] + options + ['--stdin']
```
Comment (danreeves) [inline prettier.py]: `command` isn't used here so this line can be removed

File: prettier.py [hunk 8]
```diff
@@ -58,15 +67,13 @@ def run(self, edit):
...
-                do_replace(edit, self.view, region, prettier_output)
+            prettify_code(edit, self.view, region)
+
+
+# Prettier event listeners
+class PrettierListener(sublime_plugin.EventListener):
+    def on_pre_save(self, view):
+        # Run the prettier sublime command on save if autoformat is set to true
+        autoformat = settings.get('autoformat')
+        if autoformat == True:
```
Comment (danreeves) [inline prettier.py]: should be `if autoformat is True:` according to the flake8 linter

**Review by danreeves — CHANGES_REQUESTED** [review]
Amazing PR! Seems to work perfectly, just a few little things and I'll merge it.

**Review by danreeves — APPROVED** [review]

**Comment by adriantoine** [comment]
Thanks for the review! I'm not a Python developer so I wasn't aware of the style recommendations, it's all fixed now.

**Comment by danreeves** [comment]
Not a problem. Thanks so much @adriantoine

**Comment by adriantoine** [comment]
No problem, thank you @danreeves

**Pull request merged**